File: Examples/iOS/Harmonizer/Harmonizer/Instruments/HarmonizerInstrument.m
//
//  HarmonizerInstrument.m
//  Harmonizer
//
//  Created by Aurelius Prochazka on 7/6/14.
//  Copyright (c) 2014 Aurelius Prochazka. All rights reserved.
//

#import "HarmonizerInstrument.h"
#import "AKFoundation.h"

@implementation HarmonizerInstrument

- (instancetype)init {
    self = [super init];
    if (self) {
        AKAudioInput *microphone = [[AKAudioInput alloc] init];
        [self connect:microphone];
        
        AKFFT *microphoneFFT;
        microphoneFFT = [[AKFFT alloc] initWithInput:microphone
                                             fftSize:akpi(2048)
                                             overlap:akpi(256)
                                          windowType:AKFFTWindowTypeVonHann
                                    windowFilterSize:akpi(2048)];
        [self connect:microphoneFFT];
        
        AKScaledFFT *scaledFFT;
        scaledFFT = [[AKScaledFFT alloc] initWithSignal:microphoneFFT
                                         frequencyRatio:akp(2.0)
                                    formantRetainMethod:AKScaledFFTFormantRetainMethodLifteredCepstrum
                                         amplitudeRatio:akp(2.0)
                                   cepstrumCoefficients:nil];
        [self connect:scaledFFT];
        
        AKMixedFFT *mixedFFT = [[AKMixedFFT alloc] initWithSignal1:microphoneFFT signal2:scaledFFT];
        [self connect:mixedFFT];
        
        AKResynthesizedAudio *audioOutput = [[AKResynthesizedAudio alloc] initWithSignal:mixedFFT];
        [self connect:audioOutput];
        
        
        // AUDIO OUTPUT ========================================================
        AKAudioOutput *output = [[AKAudioOutput alloc] initWithAudioSource:audioOutput];
        [self connect:output];
    }
    return self;
}

@end

File: Examples/iOS/SwiftHarmonizer/HarmonizerInstrument.swift
//
//  HarmonizerInstrument.swift
//  SwiftHarmonizer
//
//  Created by Nicholas Arner on 10/7/14.
//  Copyright (c) 2014 Aurelius Prochazka. All rights reserved.
//

class HarmonizerInstrument: AKInstrument {
    
    // INSTRUMENT DEFINITION ===================================================
    
    override init() {
        super.init()
        
        let microphone: AKAudioInput = AKAudioInput()
        connect (microphone)
        
        let microphoneFFT = AKFFT(
            input: microphone,
            fftSize: 2048.ak,
            overlap: 256.ak,
            windowType: AKFFTWindowType.VonHann,
            windowFilterSize: 2048.ak
        )
        connect(microphoneFFT)
        
        let scaledFFT = AKScaledFFT(
            signal: microphoneFFT,
            frequencyRatio: 2.ak,
            formantRetainMethod: AKScaledFFTFormantRetainMethod.LifteredCepstrum,
            amplitudeRatio: 2.ak,
            cepstrumCoefficients: nil
        )
        connect(scaledFFT)
    
        let mixedFFT = AKMixedFFT(signal1: microphoneFFT, signal2: scaledFFT)
        connect(mixedFFT)
        
        let audioOutput = AKResynthesizedAudio(signal: mixedFFT)
        connect(audioOutput)
        connect(AKAudioOutput(audioSource: audioOutput))
    }
}